### DISCUSSION TITLE
Using IAnonymousProducer without registring each producer as anonymous

### DISCUSSION THREAD

**Discussion by MehdiRifi** [Havret/dotnet-activemq-artemis-client#479] [Q&A] [answered]
Hello,
I appreciate the effort put into this project. I have a question ,is there a way to send a message to a queue using  IAnonymousProducer or anything else, without having to register the class as a typed anonymous producer? My goal is to abstract the usage of the library as much as possible.
ive tried to use the IConnection from di but it could not be resolved also.

btw I’m using the hosting and DI nugets within this api
thanks in advance.

**Comment by MehdiRifi** [comment]
I’ve found a workaround by injecting the ConnectionProvider and retrieving the connection by its name. This allows me to get a producer to send messages to the targeted queue. Are there any potential drawbacks to manually creating the producer?

```c#
    var message = new Message(messageBody);
    var connection = await _connectionProvider.GetConnectionAsync("MyAppConnection");
    var producer = await connection.CreateProducerAsync("a1", RoutingType.Anycast);
    await producer.SendAsync(message);
```

**Reply by Havret** [reply]
Hi @MehdiRifi,

It's great to hear your appreciation for the project, and thank you for reaching out with your question.

Regarding your approach to sending messages without using a typed anonymous producer, your workaround of manually creating a producer by fetching a connection from the ConnectionProvider does indeed provide a level of abstraction. However, this method might not be optimal for several reasons. As highlighted in the [documentation for the dotnet-activemq-artemis-client](https://havret.github.io/dotnet-activemq-artemis-client/docs/getting-started), connections, producers, and consumers are intended to be long-lived entities. The protocol underlying these components is optimized for persistent connections, meaning that frequently establishing new connections or producers for each message can significantly impact performance due to the additional network round trips and overhead.

Your method, which involves creating a new AMQP session and producer for every message sent, could lead to degraded performance, especially as the volume of messages increases. While this might not be an issue for infrequent message sending, it could become problematic with higher throughput.

The library's design, particularly with its dependency injection (DI) integration, aims to cache producers in a thread-safe manner. This caching ensures efficient utilization of AMQP resources, similar to how the HttpClientFactory manages HttpClient instances for HTTP communications. This design choice is intended to abstract away complexities and provide a cleaner, more efficient interface for messaging.

Given this, while your workaround could serve in specific scenarios, it's worth considering the intended use of the library's abstractions, like the strongly typed anonymous producers. These abstractions are designed to simplify messaging operations while optimizing resource management and performance.

If there are specific reasons or requirements leading you to avoid these abstractions, it might be helpful to discuss them further. There could be alternative solutions or enhancements to the library that could address your needs more effectively.

Best regards,
Havret

**Comment by MehdiRifi** [comment]
Thank you for the feedback. As the library currently only handles AMQP, we may need to switch to another protocol in the future if required. Therefore, we wanted to decouple the code from the message delivery method to make it as easy as possible to swap the implementation in the future. Could you please clarify how the producer/session lifecycle is managed in this context?

**Reply by Havret** [reply, marked as answer]
The ArtemisNetClient simplifies the AMQP API by eliminating the need for a session within the API. When you create a producer or consumer, each resource independently creates its own session, which is then retained as part of the producer or consumer objects, respectively.

As I mentioned in my previous response, AMQP resources are designed to be long-lived entities. Adhering to this best practice is relatively straightforward for console applications but can be challenging for web-based applications or applications that utilize dependency injection concepts in general. To properly use the library in these contexts, users must navigate a series of non-trivial decisions, such as how to register messaging components, whether they should be registered as scoped, transient, or singletons, when to create them (e.g., lazily upon first use for producers, or eagerly as the application starts), and in which order to close them to ensure a graceful shutdown of the application without exceptions or data loss. Mismanaging the shutdown sequence, for instance, by closing producers before consumers, could lead to exceptions if additional messages are sent during message processing. While these issues may not pose significant challenges for an experienced engineer, I have frequently observed developers handling these problems without adequate consideration or incorrectly throughout my career. To address these questions, I have developed a helper library built on top of ArtemisNetClient, named ArtemisNetClient.Extensions.DependencyInjection, which offers streamlined integration with the .NET dependency injection system.

Although I am not familiar with the specifics of your library's API, it appears that you are utilizing our DI Extensions, suggesting that your library is also built atop the .NET dependency injection system. With that in mind, I wonder why you have not considered creating an abstraction around the Anonymous Message Producer to construct your own message-sending API. Below is a conceptual draft of what this might look like:

```csharp
public static class MessagingLibraryExtensions
{
    public static void AddMessaging(this IServiceCollection services)
    {
        services.AddActiveMq(name: "my-artemis-cluster",
            endpoints: new[]
            {
                Endpoint.Create(host: "localhost", port: 5672, username: "artemis", password: "artemis"),
            })
            .AddAnonymousProducer<MyInternalMessageProducer>();

        services.AddTransient<IMessageSender, ArtemisNetClientMessageSender>();
    }
}

public class MyInternalMessageProducer
{
    private readonly IAnonymousProducer _messageSender;

    public MyInternalMessageProducer(IAnonymousProducer messageSender)
    {
        _messageSender = messageSender;
    }

    public async Task Send(string topic, string message)
    {
        await _messageSender.SendAsync(topic, RoutingType.Multicast, new Message(message));
    }
}

public interface IMessageSender
{
    Task Send(string message);
}

public class ArtemisNetClientMessageSender : IMessageSender
{
    private readonly MyInternalMessageProducer _messageProducer;

    public ArtemisNetClientMessageSender(MyInternalMessageProducer messageProducer)
    {
        _messageProducer = messageProducer;
    }

    public async Task Send(string message)
    {
        await _messageProducer.Send("mytopic", message);
    }
}
```

This approach only exposes the `IMessageSender` interface to your library users. It allows for easy substitution by providing a different internal implementation of the interface. We could further simplify this implementation by enhancing ArtemisNetClient.Extensions.DependencyInjection to allow for the registration of typed message producers as interfaces, not just concrete types. This would eliminate the need for the `MyInternalMessageProducer` class, enabling you to directly register `ArtemisNetClientMessageSender` as `IMessageSender` using the `AddAnonymousProducer` extension method:

```csharp
public static class MessagingLibraryExtensions
{
    public static void AddMessaging(this IServiceCollection services)
    {
        services.AddActiveMq(name: "my-artemis-cluster",
            endpoints: new[]
            {
                Endpoint.Create(host: "localhost", port: 5672, username: "artemis", password: "artemis"),
            })
            .AddAnonymousProducer<IMessageSender, ArtemisNetClientMessageSender>();
    }
}
```

@MehdiRifi, please let me know if this is something you would find useful, and I can add it.

Apologies for the lengthy reply, but I hope it clarifies things somewhat.

Best regards,
Havret

**Comment by MehdiRifi** [comment]
Hello **@Havret**, I truly appreciate your responsiveness and the time you've dedicated to answering my questions. Regarding the solution, I ended up implementing something quite similar to your suggestions:

1. I created a **typed anonymous producer** using the library.
2. I built an **abstraction layer** over the typed anonymous producer to hide the specific protocol and library being used.
3. For each message, I designed a **custom producer** that hides details about which queue to use and how to serialize the specific type. 

`IMessageProducer<T>  --> IMessageSender  --> TypedAnonymousProducer` (where the library is utilized).

Thank you!
